### PR TITLE
[autoscaler] make 0 default min/max workers for head node

### LINKED ISSUE #17554: [autoscaler] Default min and max workers to 0 for the head node type.
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

Most use-cases require a head node with a config distinct from the workers.
Having to explicitly state that there are no workers of head node type is confusing. This should be fixed in all cluster yamls and helm configs.

Document the change and log a warning if necessary, as, strictly speaking, this change is backwards-incompatible.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by sasha-s**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
It's customary to add a tag to the title to aid in release tracking -- added "[autoscaler]".

File: python/ray/tests/test_autoscaler.py [hunk 1]
```diff
@@ -2857,6 +2855,15 @@ def metrics_incremented():
         self.waitFor(
             metrics_incremented, fail_msg="Expected metrics to update")
 
+    def testValidateDefaultConfigi2(self):
```
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler.py:2858]: typo in function name

File: python/ray/autoscaler/_private/util.py [hunk 2]
```diff
@@ -257,26 +257,20 @@ def merge_setup_commands(config):
...
-        # Log a warning if head node type's max_workers is absent.
-        if (node_type_name == config["head_node_type"]
-                and "max_workers" not in node_type_data):
-            cli_logger.warning(
-                HEAD_TYPE_MAX_WORKERS_WARN_TEMPLATE.format(
-                    node_type=node_type_name,
-                    max_workers=config["max_workers"],
-                    version=ray.__version__))
+        if node_type_name == config["head_node_type"]:
+            node_type_data.setdefault("min_workers", 0)
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/util.py:272]: We want `max_workers` for the head node type defaulted to 0 (and to global max_workers for the rest of the node types).

I think the autoscaler infers min_workers internally 0 if min_workers is missing for a node type.

**Comment by DmitriGekhtman** [comment]
Other things that need to be changed: 

There's a test in `test_cli.py` that explicitly checks for the max_workers warning message that's been deleted -- that test and the file it uses should be deleted.

The documentation on node type max workers should be updated 
https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/doc/source/cluster/config.rst

The helm chart template https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/charts/ray/templates/raycluster.yaml#L22-L23
needs to be updated to ignore min_workers and max_workers if they're not specified in values.yaml https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/charts/ray/values.yaml#L16-L17.

There's another example that needs to be modified here
https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/components/example_cluster.yaml#L33-L35

File: python/ray/autoscaler/_private/util.py (hunk 2 above)
Comment (sasha-s) [reply python/ray/autoscaler/_private/util.py:272]: Changed it to 0 form the head node, setting min_workers to 0 as default.

File: python/ray/tests/test_autoscaler.py (hunk 1 above)
Comment (sasha-s) [reply python/ray/tests/test_autoscaler.py:2858]: thanks

**Comment by sasha-s** [comment]
> Other things that need to be changed:
> 
> There's a test in `test_cli.py` that explicitly checks for the max_workers warning message that's been deleted -- that test and the file it uses should be deleted.
> 
> The documentation on node type max workers should be updated
> https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/doc/source/cluster/config.rst
> 
> The helm chart template
> 
> https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/charts/ray/templates/raycluster.yaml#L22-L23
> 
> 
> needs to be updated to ignore min_workers and max_workers if they're not specified in values.yaml
> https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/charts/ray/values.yaml#L16-L17
> 
> .
> There's another example that needs to be modified here
> https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/deploy/components/example_cluster.yaml#L33-L35

thanks, I missed those

**Comment by DmitriGekhtman** [comment]
There's an instance of min and max workers to be removed here
https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/doc/source/cluster/config.rst#available_node_types

This spot should mention the default behavior for the head node type max workers
https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/doc/source/cluster/config.rst#available_node_typesnode_type_namenode_typemax_workers

**Comment by sasha-s** [comment]
> This spot should mention the default behavior for the head node type max workers
> https://github.com/ray-project/ray/blob/55680a1f9ed3d8d1d4762039221b9fba4645383b/doc/source/cluster/config.rst#available_node_typesnode_type_namenode_typemax_workers

done

**Comment by DmitriGekhtman** [comment]
lgtm -- I'm just going to run some quick manual checks that k8s stuff works

File: python/ray/autoscaler/_private/util.py [hunk 3]
```diff
@@ -257,27 +257,23 @@ def merge_setup_commands(config):
     return config
 
 
-def fill_node_type_max_workers(config):
+def fill_node_type_min_max_workers(config):
     """Sets default per-node max workers to global max_workers.
     This equivalent to setting the default per-node max workers to infinity,
     with the only upper constraint coming from the global max_workers.
+    Sets default per-node max workers to zero.
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/util.py:264]:
```suggestion
    Sets default per-node max workers to global max_workers.
```

File: python/ray/autoscaler/_private/util.py [hunk 4]
```diff
@@ -257,27 +257,23 @@ def merge_setup_commands(config):
...
-            cli_logger.warning(
-                HEAD_TYPE_MAX_WORKERS_WARN_TEMPLATE.format(
-                    node_type=node_type_name,
-                    max_workers=config["max_workers"],
-                    version=ray.__version__))
-
-        # The key part of this function:
-        node_type_data.setdefault("max_workers", config["max_workers"])
+        node_type_data.setdefault("min_workers", 0)
+        if node_type_name == config["head_node_type"]:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/util.py:273]: Can you please print `log.info("...")` saying that setting max_workers of head node type to 0 and a similar one for workers?

File: python/ray/autoscaler/_private/util.py (hunk 3 above)
Comment (sasha-s) [reply python/ray/autoscaler/_private/util.py:264]: We have. comment about max workers above.
Changes this one to talk about min workers.

Comment (AmeerHajAli) [reply python/ray/autoscaler/_private/util.py:264]: my main concern was that it says "Sets default per-node max workers to zero." but it actually sets it to global max workers. Maybe you meant min_workers to zero?

File: python/ray/autoscaler/_private/util.py (hunk 4 above)
Comment (sasha-s) [reply python/ray/autoscaler/_private/util.py]: done

**Comment by DmitriGekhtman** [comment]
The somewhat unorthodox convention we use for Ray github is to add reviewers as _assignees_ (makes searching PRs to be reviewed easier for reviewers)

File: python/ray/autoscaler/_private/util.py [hunk 5]
```diff
@@ -257,27 +257,28 @@ def merge_setup_commands(config):
...
-                    node_type=node_type_name,
-                    max_workers=config["max_workers"],
-                    version=ray.__version__))
-
-        # The key part of this function:
-        node_type_data.setdefault("max_workers", config["max_workers"])
+        node_type_data.setdefault("min_workers", 0)
+        if "max_workers" not in node_type_data:
+            if node_type_name == config["head_node_type"]:
+                logger.info("setting max workers for head node to 0")
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/util.py:275]:
```suggestion
                logger.info("setting max workers for head node type to 0")
```

**Comment by DmitriGekhtman** [comment]
LGTM -- would you mind doing a manual check by launching a cluster on AWS with these changes checked out locally?

**Review by DmitriGekhtman — APPROVED** [review]

**Comment by DmitriGekhtman** [comment]
Looks like there's a Mac build failure. Could you rebase or merge master to retry that?

**Comment by DmitriGekhtman** [comment]
eh, sry, realized there's one more helm issue with the helm chart.
As is, there's an incompatibility with the default `rayproject/ray:latest` operator image, since the max/min-workers-filling logic runs in the operator container.

To avoid the incompatibility, could you modify [this part](https://github.com/ray-project/ray/blob/6bfbc59f2edd544597c7e95dde785e8152edba9f/deploy/charts/ray/templates/raycluster.yaml#L22-L27) to default to 0 if minWorkers/maxWorkers is not present? There's Helm syntax for [defaults](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-default-function) which simplifies this.

**Comment by sasha-s** [comment]
> eh, sry, realized there's one more issue with the helm chart.
> 
> As is, there's an incompatibility with the default `rayproject/ray:latest` operator image, since the max/min-workers-filling logic runs in the operator container.
> 
> To avoid the incompatibility, could you modify [this part](https://github.com/ray-project/ray/blob/6bfbc59f2edd544597c7e95dde785e8152edba9f/deploy/charts/ray/templates/raycluster.yaml#L22-L27) to default to 0 if minWorkers/maxWorkers is not present? There's Helm syntax for [defaults](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-default-function) which simplifies this.



> eh, sry, realized there's one more issue with the helm chart.
> 
> As is, there's an incompatibility with the default `rayproject/ray:latest` operator image, since the max/min-workers-filling logic runs in the operator container.
> 
> To avoid the incompatibility, could you modify [this part](https://github.com/ray-project/ray/blob/6bfbc59f2edd544597c7e95dde785e8152edba9f/deploy/charts/ray/templates/raycluster.yaml#L22-L27) to default to 0 if minWorkers/maxWorkers is not present? There's Helm syntax for [defaults](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-default-function) which simplifies this.
I guess it is not feasible to remove the defaults from `rayproject/ray:latest`.
Changes min/max to default 0.

File: deploy/charts/ray/templates/raycluster.yaml [hunk 6]
```diff
@@ -19,9 +19,8 @@ spec:
   podTypes:
     {{- range $key, $val := .Values.podTypes }}
     - name: {{ $key }}
-      minWorkers: {{ $val.minWorkers }}
-      maxWorkers: {{ $val.maxWorkers }}
-      {{- if $val.rayResources }}
```
[the remark this answers is not in the document]
Comment (sasha-s) [reply deploy/charts/ray/templates/raycluster.yaml:24]: done

**Pull request merged**